Diagnostic.py

```diff
@@ -1,332 +1,352 @@
+
+        dL2dphi   = grad(self.L_G(x_fake1,self.D,self.device),\
+                         self.phi,create_graph=True,retain_graph=True)
+        
+        for ((name,_),g) in zip(self.G.named_parameters(),dL2dphi):
+            grad_gen_epoch[name]+=g.flatten()*len(x_real)
+
+        n_data +=len(x_real)
+
+      grad_D = []
+      grad_G = []
+
+      for name,_ in self.D.named_parameters():
+        grad_D.append(grad_dis_epoch[name]/n_data)
+
+      for name,_ in self.G.named_parameters():
+        grad_G.append(grad_gen_epoch[name]/n_data)
+
+    return grad_D, self.theta, grad_G, self.phi
+  
+  
+  
+class Jacobian(linalg.LinearOperator):
+  def __init__(self,first_grad,params,device,transpose=False):
+
+    self.first_grad = first_grad
+    self.params     = params
+    self.transpose  = transpose
+    self.device     = device 
+    self.dtype      = np.dtype('Float32')
+    
+    n_params = 0
+    for p in self.params:
+      n_params+=p.numel()
+
+    self.shape = (n_params,n_params)
+  
+  @staticmethod
+  def vectorize(x):
+    return torch.cat([torch.flatten(y) for y in x]).unsqueeze(-1)
+  
+  def JVP(self,v):
+    # operator J(first_grad,params)v --> Hessian(params)v
+    dF   = Jacobian.vectorize(self.first_grad)
+    u    = torch.ones_like(dF,requires_grad=True)       
+    g_v  = grad(dF,self.params,u,create_graph=True,retain_graph=True)
+    g_v  = Jacobian.vectorize(g_v)
+    ans, = grad(g_v,u,v)
+    return ans
+
+  def JTVP(self,v):
+    # operator J^T(first_grad,params)v --> Hessian^T(params)v
+    return Jacobian.vectorize(grad(Jacobian.vectorize(self.first_grad),self.params,v,retain_graph=True))
+
+  def _matvec(self,v):
+    v = np.expand_dims(v,axis=-1)
+    v = torch.tensor(v,device=self.device)
+    if self.transpose == True:
+      return self.JTVP(v).cpu().detach().numpy()
+    else:
+      return self.JVP(v).cpu().detach().numpy()
+
+  def trace(self,n_iter): ### n_iter for Hutchinson Stochastic Trace Estimators
+    trace_list = []
+    for _ in range(n_iter):
+      v = self.rand_vec()
+      trace_list.append(torch.matmul(v.T,self.JTVP(v)))
+    return torch.mean(torch.tensor(trace_list))
+  
+  def eigen_pair(self,n_eigen,which):  #### returns [eig_val, eig_vec]
+    
+    #### WHICH ################
+    #‘LM’ : largest magnitude
+    #‘SM’ : smallest magnitude
+    #‘LR’ : largest real part
+    #‘SR’ : smallest real part
+    #‘LI’ : largest imaginary part
+    #‘SI’ : smallest imaginary part
+    
+    if self.transpose == True:
+      return eigsh(self,which = which,k=n_eigen)
+    else:
+      return eigs(self,which = which,k=n_eigen)
+
+  '''random vector from Rademacher distribution'''
+
+  def rand_vec(self):
+    v = torch.cat([torch.randint_like(p.data, high=2,device=self.device).flatten()\
+          for p in self.params]).unsqueeze(-1)
+    v[v==0.]=-1. 
+    return v/np.sqrt(len(v))
+
+  '''Orthogonilise: Gram-Schmidt process '''
+
+  @staticmethod
+  def ort(w,list1):
```

Comment:
Put docstrings into their respective function to increase readability
```
  @staticmethod
  def ort(w,list1):
    '''Orthogonilise: Gram-Schmidt process '''
```

Reply:
Notations fixed